cmd/cluster/scale/defaulting/scaling.go

```diff
@@ -0,0 +1,131 @@
+	currentScalingMax        int64
+	currentScalingMin        int64
+	desiredNumWorkers        int64
+	desiredNumWorkersChanged bool
+	desiredScalingMax        int64
+	desiredScalingMaxChanged bool
+	desiredScalingMin        int64
+	desiredScalingMinChanged bool
+}
+
+func NewScaling(config ScalingConfig) (*Scaling, error) {
+	if config.AutoScalingEnabled == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.AutoScalingEnabled must not be empty", config)
+	}
+
+	s := &Scaling{
+		autoScalingEnabled:       *config.AutoScalingEnabled,
+		currentScalingMax:        *config.CurrentScalingMax,
+		currentScalingMin:        *config.CurrentScalingMin,
+		desiredNumWorkers:        *config.DesiredNumWorkers,
+		desiredNumWorkersChanged: *config.DesiredNumWorkersChanged,
+		desiredScalingMax:        *config.DesiredScalingMax,
+		desiredScalingMaxChanged: *config.DesiredScalingMaxChanged,
+		desiredScalingMin:        *config.DesiredScalingMin,
+		desiredScalingMinChanged: *config.DesiredScalingMinChanged,
+	}
+
+	return s, nil
+}
+
+func (s *Scaling) Default(ctx context.Context, scaling request.Scaling) request.Scaling {
+	rules := []ruleengine.DefaultingRule{
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return s.desiredScalingMinChanged },
+			},
+			Defaulting: func() {
+				scaling.Min = s.desiredScalingMin
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return s.desiredScalingMaxChanged },
+			},
+			Defaulting: func() {
+				scaling.Max = s.desiredScalingMax
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return !s.desiredScalingMinChanged },
+			},
+			Defaulting: func() {
+				scaling.Min = s.currentScalingMin
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return !s.desiredScalingMaxChanged },
+			},
+			Defaulting: func() {
+				scaling.Max = s.currentScalingMax
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return s.autoScalingEnabled },
+				func() bool { return !s.desiredScalingMinChanged },
+				func() bool { return !s.desiredScalingMaxChanged },
+				func() bool { return s.desiredNumWorkersChanged },
+			},
+			Defaulting: func() {
+				scaling.Min = s.desiredNumWorkers
+				scaling.Max = s.desiredNumWorkers
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return !s.autoScalingEnabled },
+				func() bool { return s.desiredScalingMinChanged },
+			},
+			Defaulting: func() {
+				scaling.Min = s.desiredScalingMin
+				scaling.Max = s.desiredScalingMin
+			},
+		},
+		{
+			Conditions: []func() bool{
+				func() bool { return !s.autoScalingEnabled },
+				func() bool { return s.desiredScalingMaxChanged },
+			},
+			Defaulting: func() {
+				scaling.Min = s.desiredScalingMax
+				scaling.Max = s.desiredScalingMax
+			},
+		},
```

Comment:
Does validation catch situation where both scaling-min or scaling-max and number of workers is given? Otherwise if I understand this correctly, this defaulting makes the cluster then pinned at number of workers and that is not always what the user intended. The user could e.g. misunderstand how this works and try to set scaling limits and then set present "desired number of workers" to some specific values - which is impossible.

Reply:
Same as above, I do not know. When there is a case you have in mind let's test and verify. There are a couple of confirmations so I am not sure how misleading all of this can be at the end.